### DISCUSSION TITLE
how to make sure SmartCacheDataset will be shuffling data

### DISCUSSION THREAD

**Discussion by etienne87** [Project-MONAI/MONAI#6967] [General]
I would like to combine a PersistentDataset and the SmartCacheDataset in order to take multiple crops from the same exam. It seems that when I feed SmartCacheDataset with a PersistentDataset (or any Dataset) it does not manage to shuffle it because this is not a sequence.

```you are shuffling a 'PersistentDataset' object which is not a subclass of 'Sequence'; `shuffle` is not guaranteed to behave correctly. E.g., non-numpy array/tensor objects with view semantics may contain duplicates after shuffling.
input data can't be shuffled in SmartCacheDataset with numpy.random.shuffle(): 'PersistentDataset' object does not support item assignment.```

What is the right way to do this?

**Comment by KumoLiu** [comment]
Hi, @etienne87, thanks for your interest here.
I didn't get your point about why you want to combine a PersistentDataset and the SmartCacheDataset. Can you just use the `transform` in `SmartCacheDataset` to crop your data? 
Thanks!

**Comment by etienne87** [comment]
Hi, here is the pipeline of data that i want to set in place:
- RawData is Preprocessed and saved in pickle with PersistentDataset (the preprocessing happens in a transform that loads the data, resize, etc.)
- Pickled Dataset is then gradually loaded by the SmartCacheDataset (with cache_num loaded in RAM, the rest is loaded on next epochs).
- The data is then streamed to a PatchDataset to take multiple patches per volume, put them into a buffer, shuffle it so that batches do not contain data from only 1 exam or 2 but multiple (ideally there is X exams for X patches)/ 
- The PatchDataset finally goes to monai DataLoader

**Comment by wyli** [comment]
maybe the data shuffling in SmartCacheDataset is not necessary
https://github.com/Project-MONAI/MONAI/blob/392c5c1b860c0f0cfd0aa14e9d4b342c8b5ef5e7/monai/data/dataset.py#L1082
we can shuffle the list of replacing indices `list(range(self._replace_num))` instead? cc @Nic-Ma 
https://github.com/Project-MONAI/MONAI/blob/392c5c1b860c0f0cfd0aa14e9d4b342c8b5ef5e7/monai/data/dataset.py#L1199
